internal/fnruntime/nodejs.go

```diff
@@ -0,0 +1,604 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fnruntime
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+)
+
+var writeJsFile sync.Once
+
+const (
+	WasmPathEnv = "KPT_FN_WASM_PATH"
+)
+
+type WasmNodejsFn struct {
+	NodeJsRunner *ExecFn
+}
+
+func NewNodejsFn(wasm string) (*WasmNodejsFn, error) {
+	var err error
+	cacheDir := filepath.Join(os.TempDir(), "kpt-wasm-fn")
+	jsPath := filepath.Join(cacheDir, "kpt-fn-wasm-glue-runner.js")
+	writeJsFile.Do(func() {
+		if _, err = os.Stat(cacheDir); os.IsNotExist(err) {
+			if err = os.Mkdir(cacheDir, 0755); err != nil {
+				return
+			}
+		}
+		if err = os.WriteFile(jsPath, glueJSCode, 0666); err != nil {
+			return
+		}
+	})
+	if err != nil {
+		return nil, fmt.Errorf("unable to create the js glue file")
+	}
+
+	wasmFile := wasm
+	if !filepath.IsAbs(wasm) {
+		wasmFileReadCloser, err := loadWasmFromImage(wasm)
+		if err != nil {
+			return nil, fmt.Errorf("unable to load wasm content from image %v: %w", wasm, err)
+		}
+		defer wasmFileReadCloser.Close()
+		// Create a module out of the Wasm file.
+		data, err := io.ReadAll(wasmFileReadCloser)
+		if err != nil {
+			return nil, fmt.Errorf("unable to read wasm content from reader: %w", err)
+		}
+		wasmFile = filepath.Join(cacheDir, "fn.wasm")
+		err = os.WriteFile(wasmFile, data, 0666)
+		if err != nil {
+			return nil, fmt.Errorf("unable to wasm content to %v: %w", wasmFile, err)
+		}
+	}
+
+	f := &WasmNodejsFn{
+		NodeJsRunner: &ExecFn{
+			Path: "node",
+			Args: []string{jsPath},
+			Env: map[string]string{
+				WasmPathEnv: wasmFile,
+			},
+		},
+	}
+	return f, nil
+}
+
+func (f *WasmNodejsFn) Run(r io.Reader, w io.Writer) error {
+	return f.NodeJsRunner.Run(r, w)
+}
+
+var glueJSCode = []byte(`const fs = require('fs');
```

Comment:
nit: maybe this var belongs in its own file since it's long and is actually its own code

Reply:
Done